### ISSUE TITLE
horovod run with pytorch1.6 torch.cuda.amp  will cause nan

### ISSUE THREAD

**Issue by roysun2006** [horovod/horovod#2175]
**Environment:**
1. Framework: PyTorch 
2. Framework version: 1.6
3. Horovod version: 0.19.5
4. MPI version: 4.0.0
5. CUDA version: 10.1
6. NCCL version: 2.4.8
7. Python version: 3.7.7
8. Spark / PySpark version:
9. OS and version: ubuntu 18.04
10. GCC version: 7.4.0

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
    yes, I did.
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?
   yes, I did

**Bug report:**
When I use the pytorch1.6 new feature torch.cuda.amp scaler and autocast, after 
3 epochs, the loss will be nan forever. 
I’d appreciate if anybody can help me! Thanks in advance!

Here is my sample code:
for _, datas in enumerate(train_loader):
        optimizer.zero_grad()
        x = tuple(to_device(model, arr.squeeze(0)) for arr in datas)

        with autocast(enabled=True):
            loss = model(*x)
            loss = loss.mean()

        scaler.scale(loss).backward()  # Backprop

        optimizer.synchronize() 
        with optimizer.skip_synchronize():
            scaler.step(optimizer)
            scaler.update()

        if hvd.rank() == 0:
            with open(args.outdir+'/train_loss.txt', 'a') as f:
                f.write(f"Epoch {epoch}, Iteration: {iteration}, Loss: {float(loss)}\n")

here is some log of the abnormal loss
Epoch 4, Iteration: 28775, Loss: nan
Epoch 4, Iteration: 28776, Loss: nan
Epoch 4, Iteration: 28777, Loss: nan
Epoch 4, Iteration: 28778, Loss: nan
Epoch 4, Iteration: 28779, Loss: nan
Epoch 4, Iteration: 28780, Loss: nan
Epoch 4, Iteration: 28781, Loss: nan
Epoch 4, Iteration: 28782, Loss: nan
Epoch 4, Iteration: 28783, Loss: nan
Epoch 4, Iteration: 28784, Loss: nan
Epoch 4, Iteration: 28785, Loss: nan
Epoch 4, Iteration: 28786, Loss: nan
Epoch 4, Iteration: 28787, Loss: nan
Epoch 4, Iteration: 28788, Loss: nan
Epoch 4, Iteration: 28789, Loss: nan
Epoch 4, Iteration: 28790, Loss: nan
Epoch 4, Iteration: 28791, Loss: nan
Epoch 4, Iteration: 28792, Loss: nan
Epoch 4, Iteration: 28793, Loss: nan
Epoch 4, Iteration: 28794, Loss: nan
Epoch 4, Iteration: 28795, Loss: nan
Epoch 4, Iteration: 28796, Loss: nan
Epoch 4, Iteration: 28797, Loss: nan
Epoch 4, Iteration: 28798, Loss: nan
Epoch 4, Iteration: 28799, Loss: nan
Epoch 4, Iteration: 28800, Loss: nan
Epoch 4, Iteration: 28801, Loss: nan
Epoch 4, Iteration: 28802, Loss: nan
Epoch 4, Iteration: 28803, Loss: nan
Epoch 4, Iteration: 28804, Loss: nan
Epoch 4, Iteration: 28805, Loss: nan
Epoch 4, Iteration: 28806, Loss: nan
Epoch 4, Iteration: 28807, Loss: nan
Epoch 4, Iteration: 28808, Loss: nan
Epoch 4, Iteration: 28809, Loss: nan
Epoch 4, Iteration: 28810, Loss: nan
Epoch 4, Iteration: 28811, Loss: nan
Epoch 4, Iteration: 28812, Loss: nan
Epoch 4, Iteration: 28813, Loss: nan

**Comment by tgaddair** [comment]
Hey @roysun2006, thanks for reporting this issue.  

@romerojosh, do you think #1949 may be able to help in this case?  If so, can you provide guidance on how to make use of it here?

**Comment by tgaddair** [comment]
@roysun2006, are you using `hvd.Compression.fp16` or just AMP without fp16 compression?

**Comment by roysun2006** [comment]
> @roysun2006, are you using `hvd.Compression.fp16` or just AMP without fp16 compression?

Thanks for your reply, I tried both hvd.Compression.fp16 and hvd.Compression.none , they got the same result

**Comment by ydcjeff** [comment]
Any updates on this? I am facing that too

**Comment by tgaddair** [comment]
Hey @ydcjeff, I'll try and take a look at this this week.

**Comment by tgaddair** [comment]
Hey @ydcjeff, I tried reproducing this with PyTorch Lightning + Horovod + PyTorch 1.6 (CUDA 10.1), but no luck.

Here was my Trainer config:

```
trainer = pl.Trainer(gpus=1, distributed_backend='horovod', precision=16)
```

Can you share a simple repro?

**Comment by ydcjeff** [comment]
Hey @tgaddair thank you for your answer. I will share a minimal reproduction. Can you try with `gpus=2` in the meantime?
I can confirm that horovod works with 1 gpu and torch.cuda.amp, but 2 gpus doesn't

**Comment by tgaddair** [comment]
Hey @ydcjeff, when using Horovod, you should always set `gpus=1`.  If you want to run with two GPUs, you can do so through `horovodrun`:

```
horovodrun -np 2 python train.py
```

Can you try that with `gpus=1` and let me know if it works?

**Comment by thecooltechguy** [comment]
Hi, I'm also facing this same issue

**Comment by tgaddair** [comment]
Hey @thecooltechguy, do you have a minimal repro script I can run to debug further?

**Comment by turangojayev** [comment]
Having exactly the same problem, amp with ddp works, but not with horovod. Don't really have a repro script, however ddp seems to not update the weights in case of `nan`s in the grad, whereas horovod updates the weights and hence, loss keeps getting `nan` all the time after that point.

**Comment by tgaddair** [comment]
Thanks for the additional context, @turangojayev! Let me sync with some AMP folks to see if we can get to the bottom of this.

**Comment by kehuanfeng** [comment]
I met the same issue with horovod 0.21.3. Is there any updating on this?
